flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -510,32 +510,34 @@ private void stopDispatcherServices() throws Exception {
 
     @Override
     public CompletableFuture<Acknowledge> submitJob(JobGraph jobGraph, Time timeout) {
-        log.info(
-                "Received JobGraph submission '{}' ({}).", jobGraph.getName(), jobGraph.getJobID());
+        final JobID jobID = jobGraph.getJobID();
+        log.info("Received JobGraph submission '{}' ({}).", jobGraph.getName(), jobID);
 
         try {
-            if (isDuplicateJob(jobGraph.getJobID())) {
-                if (isInGloballyTerminalState(jobGraph.getJobID())) {
+            final boolean inGloballyTerminalState = isInGloballyTerminalState(jobID);
+            if (inGloballyTerminalState
+                    || jobManagerRunnerRegistry.isRegistered(jobID)
+                    || submittedAndWaitingTerminationJobIDs.contains(jobID)) {
+                if (inGloballyTerminalState) {
                     log.warn(
                             "Ignoring JobGraph submission '{}' ({}) because the job already reached a globally-terminal state (i.e. {}) in a previous execution.",
                             jobGraph.getName(),
-                            jobGraph.getJobID(),
+                            jobID,
                             Arrays.stream(JobStatus.values())
                                     .filter(JobStatus::isGloballyTerminalState)
                                     .map(JobStatus::name)
                                     .collect(Collectors.joining(", ")));
                 }
 
                 final DuplicateJobSubmissionException exception =
-                        isInGloballyTerminalState(jobGraph.getJobID())
-                                ? DuplicateJobSubmissionException.ofGloballyTerminated(
-                                        jobGraph.getJobID())
-                                : DuplicateJobSubmissionException.of(jobGraph.getJobID());
+                        inGloballyTerminalState
+                                ? DuplicateJobSubmissionException.ofGloballyTerminated(jobID)
+                                : DuplicateJobSubmissionException.of(jobID);
                 return FutureUtils.completedExceptionally(exception);
```

Comment:
What about the following:
```
            if (isInGloballyTerminalState(jobID)) {
                log.warn(
                        "Ignoring JobGraph submission '{}' ({}) because the job already reached a globally-terminal state (i.e. {}) in a previous execution.",
                        jobGraph.getName(),
                        jobID,
                        Arrays.stream(JobStatus.values())
                                .filter(JobStatus::isGloballyTerminalState)
                                .map(JobStatus::name)
                                .collect(Collectors.joining(", ")));

                return FutureUtils.completedExceptionally(
                        DuplicateJobSubmissionException.ofGloballyTerminated(jobID));
            } else if (jobManagerRunnerRegistry.isRegistered(jobID)
                    || submittedAndWaitingTerminationJobIDs.contains(jobID)) {
                // job with the given jobID is not terminated, yet
                return FutureUtils.completedExceptionally(
                        DuplicateJobSubmissionException.of(jobID));
```
I just noticed that we could improve the if condition. That would require only a single method call. WDYT?